dsp/modules/snowflake.py

```diff
@@ -0,0 +1,142 @@
+"""Module for interacting with Snowflake Cortex."""
+import json
+from typing import Any
+
+import backoff
+from pydantic_core import PydanticCustomError
+
+from dsp.modules.lm import LM
+
+try:
+    from snowflake.snowpark import Session
+    from snowflake.snowpark import functions as snow_func
+
+except ImportError:
+    pass
+
+
+def backoff_hdlr(details) -> None:
+    """Handler from https://pypi.org/project/backoff ."""
+    print(
+        f"Backing off {details['wait']:0.1f} seconds after {details['tries']} tries ",
+        f"calling function {details['target']} with kwargs",
+        f"{details['kwargs']}",
+    )
+
+
+def giveup_hdlr(details) -> bool:
+    """Wrapper function that decides when to give up on retry."""
+    if "rate limits" in str(details):
+        return False
+    return True
+
+
+class Snowflake(LM):
+    """Wrapper around Snowflake's CortexAPI.
+
+    Currently supported models include 'mistral-large','reka-flash','mixtral-8x7b',
+    'llama2-70b-chat','mistral-7b','gemma-7b'.
+    """
+
+    def __init__(self, model: str = "mistral-large", credentials=None, **kwargs):
+        """Parameters
+
+        ----------
+        model : str
+            Which pre-trained model from Snowflake to use?
+            Choices are 'mistral-large','reka-flash','mixtral-8x7b','llama2-70b-chat','mistral-7b','gemma-7b'
+            Full list of supported models is available here: https://docs.snowflake.com/en/user-guide/snowflake-cortex/llm-functions#complete
+        **kwargs: dict
+            Additional arguments to pass to the API provider.
+        """
+        super().__init__(model)
+
+        self.model = model
+        cortex_models = ["mistral-large", "reka-flash", "mixtral-8x7b", "llama2-70b-chat", "mistral-7b", "gemma-7b"]
+
+        if model in cortex_models:
+            self.available_args = {
+                "max_tokens",
+                "temperature",
+                "top_p",
+            }
+        else:
+            raise PydanticCustomError(
+                "model",
+                'model name is not valid, got "{model_name}"',
+            )
+
+        self.client = self._init_cortex(credentials=credentials)
+        self.provider = "Snowflake"
+        self.kwargs = {
+            **self.kwargs,
+            "temperature": 0.7,
+            "max_output_tokens": 1024,
+            "top_p": 1.0,
+            "top_k": 1,
+            **kwargs,
+        }
```

Comment:
need to add `self.history: list[dict[str, Any]] = []`

Reply:
Done